### PR TITLE
3620 - Implemented PUT for Memory, Storage, Processing, and Resolution

### PR DESCRIPTION
**Pull request by CallumBradburyBJSS**
Ants will do just about anything to get the sugary secretions of sap-sucking insects, such as aphids or leafhoppers. To keep the honeydew in close supply, some ants herd aphids, carrying the soft-bodied pests from plant to plant. Leafhoppers sometimes take advantage of this nurturing tendency in ants and leave their young to be raised by the ants. This allows the leafhoppers to raise another brood.

### CODE REVIEW

File: Restart Integration Env.ps1 [hunk 1]
```diff
@@ -0,0 +1,2 @@
+﻿& '.\Tear Down Environment.ps1' i
+& '.\Launch Environment.ps1' i
```
Comment (kat1906) [inline Restart Integration Env.ps1]: wheres the .sh equivalent?? 😄

File: tests/NHSD.BuyingCatalogue.API.IntegrationTests/Features/ClientApplicationTypes/Supplier Edit Client Application Type.2726.feature [hunk 2]
```diff
@@ -68,7 +68,7 @@ Scenario: 3. Client Application Types that we do not understand are ignored
...
+        | Sln1     | An full online medicine system | Online medicine 1   | { "ClientApplicationTypes" : [ "browser-based", "native-mobile" ], "BrowsersSupported": [],  "MobileResponsive": null, "Plugins": null, "MinimumConnectionSpeed": null, "MinimumDesktopResolution": null, "HardwareRequirements": null, "NativeMobileHardwareRequirements": null, "NativeDesktopHardwareRequirements": null, "AdditionalInformation": null, "MobileFirstDesign": null, "NativeMobileFirstDesign": null, "MobileOperatingSystems": null, "MobileConnectionDetails": null, "MobileMemoryAndStorage": null, "MobileThirdParty": null, "NativeMobileAdditionalInformation": null, "NativeDesktopOperatingSystemsDescription": null, "NativeDesktopMinimumConnectionSpeed": null, "NativeDesktopThirdParty": null, "NativeDesktopMemoryAndStorage": null } |
```
Comment (RuthBerry96) [inline tests/NHSD.BuyingCatalogue.API.IntegrationTests/Features/ClientApplicationTypes/Supplier Edit Client Application Type.2726.feature]: Needs correcting |

File: tests/NHSD.BuyingCatalogue.API.IntegrationTests/Features/NativeDesktop/NativeDesktopMemoryAndStorage/Supplier Edit Native Desktop MemoryAndStorage Validation.3620.feature [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        | Id    | OrganisationName |
+        | Sup 1 | GPs-R-Us         |
+    And Solutions exist
+        | SolutionID | SolutionName   | OrganisationName | SupplierStatusId | SupplierId |
+        | Sln1       | MedicOnline    | GPs-R-Us         | 1                | Sup 1      |
+
+@3620
+Scenario Outline: 1. Minimum Memory Requirement Field Fails Validation
+    Given SolutionDetail exist
+        | Solution | SummaryDescription             | FullDescription   | ClientApplication                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         |
```
Comment (RuthBerry96) [inline tests/NHSD.BuyingCatalogue.API.IntegrationTests/Features/NativeDesktop/NativeDesktopMemoryAndStorage/Supplier Edit Native Desktop MemoryAndStorage Validation.3620.feature]: For validation feature files, it's often nicer to have the client application empty, or then only contain your changes (Memory & Storage) to minimise it

File: tests/NHSD.BuyingCatalogue.API.IntegrationTests/Features/NativeDesktop/NativeDesktopMemoryAndStorage/Supplier Edit Native Desktop MemoryAndStorage Validation.3620.feature [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        | Sln1     | An full online medicine system | Online medicine 1 | { "ClientApplicationTypes": [], "BrowsersSupported": [], "MobileResponsive": null, "Plugins": null, "MinimumConnectionSpeed": null, "MinimumDesktopResolution": null, "HardwareRequirements": null, "NativeMobileHardwareRequirements": null, "NativeDesktopHardwareRequirements": null, "AdditionalInformation": null, "MobileFirstDesign": null, "NativeMobileFirstDesign": null, "MobileOperatingSystems": null, "MobileConnectionDetails": null, "MobileMemoryAndStorage": null, "MobileThirdParty": null, "NativeMobileAdditionalInformation": null, "NativeDesktopOperatingSystemsDescription": null, "NativeDesktopMinimumConnectionSpeed": null } |
+Examples:
+        | ErrorType | FieldValue                  |
+        | maxLength | A string with length of 301 |
+        | required  |                             |
+        | required  | NULL                        |
+
```
Comment (RuthBerry96) [inline tests/NHSD.BuyingCatalogue.API.IntegrationTests/Features/NativeDesktop/NativeDesktopMemoryAndStorage/Supplier Edit Native Desktop MemoryAndStorage Validation.3620.feature]: Maybe do another Scenario with multiple  invalid answers, so you can have several validation failures (line 63)

File: tests/NHSD.BuyingCatalogue.Solutions.API.UnitTests/NativeDesktop/NativeDesktopMemoryAndStorageControllerTests.cs [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+                            c.Data.MinimumCpu == request.MinimumCpu &&
+                            c.Data.RecommendedResolution == request.RecommendedResolution
+                        ),
+                        It.IsAny<CancellationToken>()), Times.Once);
+            }
+
+            [Test]
+            public async Task UpdateInvalidReturnsBadRequestWithValidationDetails()
+            {
+                _resultDictionary.Add("minimum-memory-requirement", "required");
```
Comment (RuthBerry96) [inline tests/NHSD.BuyingCatalogue.Solutions.API.UnitTests/NativeDesktop/NativeDesktopMemoryAndStorageControllerTests.cs]: Is it worth showing all cases of the validation in the controller tests?

File: tests/NHSD.BuyingCatalogue.Solutions.Application.UnitTests/Solutions/UpdateNativeDesktopMemoryAndStorageTests.cs [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+                        JToken.Parse(c.ClientApplication).SelectToken(StorageRequirementsToken).Value<string>() == _storageRequirements &&
+                        JToken.Parse(c.ClientApplication).SelectToken(MinimumCpuToken).Value<string>() == _minimumCpu &&
+                        JToken.Parse(c.ClientApplication).SelectToken(RecommendedResolutionToken).Value<string>() == _recommendedResolution
+                ), It.IsAny<CancellationToken>()), Times.Once);
+            validationResult.IsValid.Should().BeTrue();
+        }
+
+        [Test]
+        public async Task MissingDataShouldReturnRequiredValidationResult()
+        {
```
Comment (RuthBerry96) [inline tests/NHSD.BuyingCatalogue.Solutions.Application.UnitTests/Solutions/UpdateNativeDesktopMemoryAndStorageTests.cs]: Test cases?

File: tests/NHSD.BuyingCatalogue.Solutions.Application.UnitTests/Solutions/UpdateNativeDesktopMemoryAndStorageTests.cs [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+            validationResult.IsValid.Should().BeFalse();
+            validationResult.Should().BeOfType<RequiredMaxLengthResult>();
+            var reqMaxLengthResult = validationResult as RequiredMaxLengthResult;
+            reqMaxLengthResult.Required.Should().BeEquivalentTo("minimum-memory-requirement", "storage-requirements-description", "minimum-cpu");
+        }
+
+        [Test]
+        public async Task TooLongDataShouldReturnMaxLengthValidationResult()
+        {
+            _storageRequirements = new string('a', 301);
```
Comment (RuthBerry96) [inline tests/NHSD.BuyingCatalogue.Solutions.Application.UnitTests/Solutions/UpdateNativeDesktopMemoryAndStorageTests.cs]: Test Cases

File: tests/NHSD.BuyingCatalogue.Solutions.API.UnitTests/NativeDesktop/NativeDesktopMemoryAndStorageControllerTests.cs (hunk 5 above)
Comment (CallumBradburyBJSS) [reply tests/NHSD.BuyingCatalogue.Solutions.API.UnitTests/NativeDesktop/NativeDesktopMemoryAndStorageControllerTests.cs]: I don't think so, all we're verifying is that whatever is returned by Mediatr is reflected accurately in the returned result. The cases are covered by the Handler tests

**Pull request merged**